src/parser/smt2/smt2_term_parser.cpp

```diff
@@ -0,0 +1,168 @@
+{
+  Token tok = d_lex.nextToken();
+  std::string id = tokenStrToSymbol(tok);
+  // run the check
+  if (!d_state.isAbstractValue(id))
+  {
+    // if an abstract value, SolverEngine handles declaration
+    d_state.checkDeclaration(id, check, type);
+  }
+  return id;
+}
+
+std::vector<std::string> Smt2TermParser::parseSymbolList(DeclarationCheck check,
+                                                         SymbolType type)
+{
+  d_lex.eatToken(Token::LPAREN_TOK);
+  std::vector<std::string> symbols;
+  Token tok = d_lex.nextToken();
+  while (tok != Token::RPAREN_TOK)
+  {
+    d_lex.reinsertToken(tok);
+    std::string sym = parseSymbol(check, type);
+    symbols.push_back(sym);
+    tok = d_lex.nextToken();
+  }
+  return symbols;
+}
+
+std::string Smt2TermParser::parseKeyword()
+{
+  d_lex.eatToken(Token::KEYWORD);
+  std::string s = d_lex.tokenStr();
+  // strip off the initial colon
+  return s.substr(1);
+}
+
+uint32_t Smt2TermParser::parseIntegerNumeral()
+{
+  d_lex.eatToken(Token::INTEGER_LITERAL);
+  return tokenStrToUnsigned();
+}
+
+uint32_t Smt2TermParser::tokenStrToUnsigned()
+{
+  // forbid leading zeroes if in strict mode
+  if (d_lex.isStrict())
+  {
+    if (d_lex.tokenStr()[0] == '0')
+    {
+      d_lex.parseError("Numeral with leading zeroes are forbidden");
+    }
+  }
+  uint32_t result;
+  std::stringstream ss;
+  ss << d_lex.tokenStr();
+  ss >> result;
+  return result;
+}
+
+std::string Smt2TermParser::tokenStrToSymbol(Token tok)
+{
+  std::string id;
+  switch (tok)
+  {
+    case Token::SYMBOL: id = d_lex.tokenStr(); break;
+    case Token::QUOTED_SYMBOL:
+      id = d_lex.tokenStr();
+      // strip off the quotes
+      id = id.substr(1, id.size() - 2);
+      break;
+    case Token::UNTERMINATED_QUOTED_SYMBOL:
+      d_lex.parseError("Expected SMT-LIBv2 symbol", true);
+      break;
+    default:
+      d_lex.unexpectedTokenError(tok, "Expected SMT-LIBv2 symbol");
+      break;
+  }
+  return id;
+}
+
+std::vector<std::string> Smt2TermParser::parseNumeralList()
+{
+  std::vector<std::string> numerals;
+  Token tok = d_lex.nextToken();
+  while (tok == Token::INTEGER_LITERAL)
+  {
+    numerals.emplace_back(d_lex.tokenStr());
+    tok = d_lex.nextToken();
+  }
+  d_lex.reinsertToken(tok);
+  return numerals;
+}
+
+std::string Smt2TermParser::parseStr(bool unescape)
+{
+  d_lex.eatToken(Token::STRING_LITERAL);
+  std::string s = d_lex.tokenStr();
+  if (unescape)
+  {
+    unescapeString(s);
+  }
+  return s;
+}
+
+void Smt2TermParser::unescapeString(std::string& s)
+{
+  // strip off the quotes
+  s = s.substr(1, s.size() - 2);
+  for (size_t i = 0, ssize = s.size(); i < ssize; i++)
+  {
+    if ((unsigned)s[i] > 127 && !isprint(s[i]))
+    {
+      d_lex.parseError(
+          "Extended/unprintable characters are not "
+          "part of SMT-LIB, and they must be encoded "
+          "as escape sequences");
+    }
+  }
+  char* p_orig = strdup(s.c_str());
```

Comment:
Two things.
1. Would this be easier to read using the C++ string interface, instead of C char*s?
2. Is this allocation needed? Could you work in place on `s`?